### PR TITLE
[telegram] proxy support (SOCKS5 and HTTP)

### PR DESCRIPTION
**Pull request by ehorohorin**

### CODE REVIEW

**Comment by ZzetT** [comment]
Since there was another PR from me which was merged today, you had some conflicts again. But I resolved them already via the web interface. But it would be good if you can re-check them. Apart from that, could you please also run the "mvn spotless:apply" command in the telegram folder so reformat the code? That should (hopefully) lead to less formatting differences, because this is what I did with the PR recently merged to 2.5.x. Thanks!

**Comment by ehorohorin** [comment]
> Since there was another PR from me which was merged today, you had some conflicts again. But I resolved them already via the web interface. But it would be good if you can re-check them. Apart from that, could you please also run the "mvn spotless:apply" command in the telegram folder so reformat the code? That should (hopefully) lead to less formatting differences, because this is what I did with the PR recently merged to 2.5.x. Thanks!

https://github.com/ehorohorin/openhab2-addons/commit/275ef934f0a36d38e1ad745ecd3dcfa81e737072

**Comment by ZzetT** [comment]
Thanks. It's fine from my side now.
Strange, that the code formatter depends on the host system where you are executing it. It seems that when I run it on Windows I get Windows line breaks and you run it on a UNIX and got UNIX line breaks.

**Comment by bjoernbrings** [comment]
With that formatting problems it is not so easy to review it. @ehorohorin could you change that? Usually you could take a text editor and switch the linebreak back.

**Comment by ehorohorin** [comment]
> With that formatting problems it is not so easy to review it. @ehorohorin could you change that? Usually you could take a text editor and switch the linebreak back.

Let me try. Is there any mvn plugin to quickly fix this before commit? I supposed that mvn spotless will do that.

**Comment by ZzetT** [comment]
> With that formatting problems it is not so easy to review it.

I agree, I reviewed it locally where I turned of the comparison for the line breaks so that I can see the actual changes.

> I supposed that mvn spotless will do that.
Yeah, I would also expect that. Is it worth to open an issue for that?

File: bundles/org.openhab.binding.telegram/src/main/java/org/openhab/binding/telegram/internal/TelegramHandler.java [hunk 1]
```diff
@@ -1,314 +1,337 @@
...
+        String proxyHost = config.getProxyHost();
+        String proxyPort = config.getProxyPort();
+        String proxyType = config.getProxyType();
+
+        if (proxyHost != null && proxyPort != null) {
+            InetSocketAddress proxyAddr = new InetSocketAddress(proxyHost, Integer.parseInt(proxyPort));
+
+            Proxy.Type proxyTypeParam = Proxy.Type.SOCKS;
+
+            if (proxyType != null && "HTTP".equals(proxyType)) {
```
Comment (martinvw) [inline bundles/org.openhab.binding.telegram/src/main/java/org/openhab/binding/telegram/internal/TelegramHandler.java]: This null check is not needed, the benefit of putting the literal on the left is that you can skip it :-)

File: bundles/org.openhab.binding.telegram/src/main/resources/ESH-INF/thing/thing-types.xml [hunk 2]
```diff
@@ -1,85 +1,102 @@
...
+				<options>
+					<option value="">No Formatting</option>
+					<option value="HTML">HTML</option>
+					<option value="Markdown">Markdown</option>
+				</options>
+				<default></default>
+				<description>Support for formatted messages, values: Markdown or HTML. Default: no formatting is used.</description>
+			</parameter>
+			<parameter name="proxyHost" type="text" required="false" multiple="false">
+				<label>Proxy Host</label>
```
Comment (martinvw) [inline bundles/org.openhab.binding.telegram/src/main/resources/ESH-INF/thing/thing-types.xml]: Please add:

```
<context>network-address</context>
```

File: bundles/org.openhab.binding.telegram/src/main/resources/ESH-INF/thing/thing-types.xml [hunk 3]
```diff
@@ -1,85 +1,102 @@
...
+				<label>Parse Mode</label>
+				<options>
+					<option value="">No Formatting</option>
+					<option value="HTML">HTML</option>
+					<option value="Markdown">Markdown</option>
+				</options>
+				<default></default>
+				<description>Support for formatted messages, values: Markdown or HTML. Default: no formatting is used.</description>
+			</parameter>
+			<parameter name="proxyHost" type="text" required="false" multiple="false">
```
Comment (martinvw) [inline bundles/org.openhab.binding.telegram/src/main/resources/ESH-INF/thing/thing-types.xml]: false is default for multiple afaik we leave it away in other bindings.

File: bundles/org.openhab.binding.telegram/src/main/resources/ESH-INF/thing/thing-types.xml [hunk 4]
```diff
@@ -1,85 +1,102 @@
...
+				<label>Proxy Host</label>
+				<description>Enter your proxy host. It will be used for telegram binding only and doesn't affect entire system.</description>
+			</parameter>
+			<parameter name="proxyPort" type="text" required="false" multiple="false">
```
Comment (martinvw) [inline bundles/org.openhab.binding.telegram/src/main/resources/ESH-INF/thing/thing-types.xml]: I believe we tend to mark it as a number with a proper range, please check some other examples, better change the type to an Integer then as well. It will also change the example.

**Comment by martinvw** [comment]
> With that formatting problems it is not so easy to review it. @ehorohorin could you change that? Usually you could take a text editor and switch the linebreak back. 

You can ignore whitespace changes in the github view: https://github.blog/2018-05-01-ignore-white-space-in-code-review/

**Comment by ehorohorin** [comment]
Could you clarify the next step? I've made lots of changes based on your comments, should I re-create new PR?

**Comment by ZzetT** [comment]
@martinvw @bjoernbrings Do you know if there is any reason why this branch is still not merged although IMHO it's ready for almost a month now?

**Comment by bjoernbrings** [comment]
@martinvw requested changes. Until he approved the merge is blocked (at least for someone with my level of access rights)

**Comment by ehorohorin** [comment]
@bjoernbrings should I perform any additional action?

File: bundles/org.openhab.binding.telegram/src/main/java/org/openhab/binding/telegram/bot/TelegramActions.java [hunk 5]
```diff
@@ -1,507 +1,507 @@
...
+                localHandler.addMessageId(chatId, replyId, retMessage.message().messageId());
+            }
+            return true;
+        }
+        return false;
+    }
+
+    @RuleAction(label = "Telegram message", description = "Sends a Telegram via Telegram API")
+    public boolean sendTelegram(@ActionInput(name = "chatId") @Nullable Long chatId,
+            @ActionInput(name = "message") @Nullable String format,
```
Comment (J-N-K) [inline bundles/org.openhab.binding.telegram/src/main/java/org/openhab/binding/telegram/bot/TelegramActions.java]: Why did you change the parameter name?

File: bundles/org.openhab.binding.telegram/src/main/java/org/openhab/binding/telegram/internal/TelegramHandler.java [hunk 6]
```diff
@@ -1,314 +1,337 @@
...
+        }
+
+        OkHttpClient.Builder prepareConnection = new OkHttpClient.Builder().connectTimeout(75, TimeUnit.SECONDS)
+                .readTimeout(75, TimeUnit.SECONDS);
+
+        String proxyHost = config.getProxyHost();
+        Integer proxyPort = config.getProxyPort();
+        String proxyType = config.getProxyType();
+
+        if (proxyHost != null) {
```
Comment (J-N-K) [inline bundles/org.openhab.binding.telegram/src/main/java/org/openhab/binding/telegram/internal/TelegramHandler.java]:
```suggestion
        if (proxyHost != null && proxyPort !=null) {
```

there is no guarantee that `proxyPort` is non-null when `proxyHost` is non-null. If it is null you'll get a NPE in l. 159.

File: bundles/org.openhab.binding.telegram/src/main/resources/ESH-INF/thing/thing-types.xml [hunk 7]
```diff
@@ -1,85 +1,103 @@
...
+				<label>Parse Mode</label>
+				<options>
+					<option value="">No Formatting</option>
+					<option value="HTML">HTML</option>
+					<option value="Markdown">Markdown</option>
+				</options>
+				<default></default>
+				<description>Support for formatted messages, values: Markdown or HTML. Default: no formatting is used.</description>
+			</parameter>
+			<parameter name="proxyHost" type="text" required="false">
```
Comment (J-N-K) [inline bundles/org.openhab.binding.telegram/src/main/resources/ESH-INF/thing/thing-types.xml]:
```suggestion
			<parameter name="proxyHost" type="text">
```

`required=false` is the default (also applies below)

File: bundles/org.openhab.binding.telegram/src/main/java/org/openhab/binding/telegram/bot/TelegramActions.java (hunk 5 above)
Comment (J-N-K) [reply bundles/org.openhab.binding.telegram/src/main/java/org/openhab/binding/telegram/bot/TelegramActions.java]: Can you comment here?

Comment (ehorohorin) [reply bundles/org.openhab.binding.telegram/src/main/java/org/openhab/binding/telegram/bot/TelegramActions.java]: @J-N-K which parameter name was changed?

Comment (J-N-K) [reply bundles/org.openhab.binding.telegram/src/main/java/org/openhab/binding/telegram/bot/TelegramActions.java]: `message` to `format`. IMO this is not a good idea because

1) if no arguments are given, it's the message
2) the ActionInput is still named message, this is confusing

And please do not change the ActionInput name, this would be a breaking change and could not be merged to 2.5.x

Comment (J-N-K) [inline bundles/org.openhab.binding.telegram/src/main/java/org/openhab/binding/telegram/bot/TelegramActions.java]:
```suggestion
            @ActionInput(name = "message") @Nullable String message,
```

File: bundles/org.openhab.binding.telegram/src/main/java/org/openhab/binding/telegram/bot/TelegramActions.java [hunk 8]
```diff
@@ -1,507 +1,507 @@
...
+            @ActionInput(name = "args") @Nullable Object... args) {
+        return sendTelegram(chatId, String.format(format, args));
```
Comment (J-N-K) [inline bundles/org.openhab.binding.telegram/src/main/java/org/openhab/binding/telegram/bot/TelegramActions.java]:
```suggestion
        return sendTelegram(chatId, String.format(message, args));
```

File: bundles/org.openhab.binding.telegram/src/main/java/org/openhab/binding/telegram/bot/TelegramActions.java [hunk 9]
```diff
@@ -1,507 +1,507 @@
...
+    }
+
+    @RuleAction(label = "Telegram message", description = "Sends a Telegram via Telegram API")
+    public boolean sendTelegram(@ActionInput(name = "message") @Nullable String format,
```
Comment (J-N-K) [inline bundles/org.openhab.binding.telegram/src/main/java/org/openhab/binding/telegram/bot/TelegramActions.java]:
```suggestion
    public boolean sendTelegram(@ActionInput(name = "message") @Nullable String message,
```

File: bundles/org.openhab.binding.telegram/src/main/java/org/openhab/binding/telegram/bot/TelegramActions.java [hunk 10]
```diff
@@ -1,507 +1,507 @@
...
+            @ActionInput(name = "args") @Nullable Object... args) {
+        TelegramHandler localHandler = handler;
+        if (localHandler != null) {
+            for (Long chatId : localHandler.getChatIds()) {
+                if (!sendTelegram(chatId, format, args)) {
```
Comment (J-N-K) [inline bundles/org.openhab.binding.telegram/src/main/java/org/openhab/binding/telegram/bot/TelegramActions.java]:
```suggestion
                if (!sendTelegram(chatId, message, args)) {
```

**Comment by J-N-K** [comment]
DCO failed because of commited suggestions

**Comment by J-N-K** [comment]
@martinvw Are your changes addressed?

**Comment by martinvw** [comment]
@ehorohorin did you also claim the bounty?

**Comment by martinvw** [comment]
Feel free to claim the bounty on: https://www.bountysource.com/issues/57694823-telegram-add-socks5-support and sorry for the delay :-)

**Comment by martinvw** [comment]
@ehorohorin there is still a bounty which I think belongs to you feel free to claim it. And if you don't want to take it out of the system you can reinvest it. https://www.bountysource.com/issues/57694823-telegram-add-socks5-support


If you don't claim it I will reinvest it myself, but I really think it belongs to you!

**Comment by ehorohorin** [comment]
@martinvw could you please reinvest it on behalf on me? Many thanks

**Comment by martinvw** [comment]
> @martinvw could you please reinvest it on behalf on me? Many thanks

Done I have the idea many people would like an openHAB 2 / 3 version of the HTTP binding and put the bounty on that.
